apps/rush-lib/src/cli/utilities/pnpm/PnpmLinkManager.ts

```diff
@@ -0,0 +1,221 @@
+    return new Promise<void>((resolve: () => void, reject: (reason: Error) => void): void => {
+      // go ahead and find out the registry we are using, we need this to locate the pnpm store folder
+      const registryUrl: string = Utilities.executeCommandAndCaptureOutput(
+        this._rushConfiguration.packageManagerToolFilename,
+        ['config', 'get', 'registry'],
+        this._rushConfiguration.commonTempFolder,
+        process.env);
+
+      const encodedRegistry: string = encodeRegistry(registryUrl);
+
+      const registryPath: string = path.join(
+        this._rushConfiguration.commonTempFolder,
+        RushConstants.nodeModulesFolderName,
+        `.${encodedRegistry}`
+      );
+
+      try {
+        const rushLinkJson: IRushLinkJson = { localLinks: {} };
+
+        for (const rushProject of this._rushConfiguration.projects) {
+          console.log(os.EOL + 'LINKING: ' + rushProject.packageName);
+          this._linkProject(registryPath, rushProject, rushLinkJson);
+        }
+
+        console.log(`Writing "${this._rushConfiguration.rushLinkJsonFilename}"`);
+        JsonFile.save(rushLinkJson, this._rushConfiguration.rushLinkJsonFilename);
+
+        resolve();
+      } catch (error) {
+        reject(error);
+      }
+    });
+  }
+
+  /**
+   * This is called once for each local project from Rush.json.
+   * @param project             The local project that we will create symlinks for
+   * @param rushLinkJson        The common/temp/rush-link.json output file
+   */
+  private _linkProject(
+    pnpmStorePath: string,
+    project: RushConfigurationProject,
+    rushLinkJson: IRushLinkJson): void {
+
+    // Naively, PNPM creates a directed acyclic graph, rather than a tree
+    // thus if sp-http and sp-core-library both depend on GCB@1.2.3,
+    // and GCB@1.2.3 depends on gulp-typescript, then for both cases it must
+    // be the exact same version of gulp-typescript
+    // thus, we only need to link the direct dependencies of this project to the
+    // common folder
+
+    // Example: "project1"
+    const unscopedTempProjectName: string = Utilities.parseScopedPackageName(project.tempProjectName).name;
+
+    // Example: "C:\MyRepo\common\temp\projects\project1
+    const extractedFolder: string = path.join(this._rushConfiguration.commonTempFolder,
+      RushConstants.rushTempProjectsFolderName, unscopedTempProjectName);
+
+    // Example: "C:\MyRepo\common\temp\projects\project1.tgz"
+    const tarballFile: string = path.join(this._rushConfiguration.commonTempFolder,
+      RushConstants.rushTempProjectsFolderName, unscopedTempProjectName + '.tgz');
+
+    // Example: "C:\MyRepo\common\temp\projects\project1\package.json"
+    const packageJsonFilename: string = path.join(extractedFolder, 'package', 'package.json');
+
+    // @todo this code is duplicated
+    Utilities.createFolderWithRetry(extractedFolder);
+    tar.extract({
+      cwd: extractedFolder,
+      file: tarballFile,
+      sync: true
+    });
```

Comment:
Why are we unzipping tarballs?

This is a longer conversation.  Let's discuss the PNPM-specific subclasses in a separate PR.

Reply:
Alright

---
In reply to: [151794771](https://github.com/Microsoft/web-build-tools/pull/427#discussion_r151794771) [](ancestors = 151794771)

Reply:
Now that we have made the other change in Rush which leaves the original package.json files hanging around, we probably don't need to actually extract the tarball

---
In reply to: [152091975](https://github.com/Microsoft/web-build-tools/pull/427#discussion_r152091975) [](ancestors = 152091975,151794771)